torch/onnx/_internal/fx/_pass.py

```diff
@@ -0,0 +1,204 @@
+        -        view = input_ids.view(-1, 3);  input_ids = None
+        +        view: i64[1, 3] = torch.ops.aten.view.default(input_ids, [-1, 3]);  input_ids = None
+    """
+
+    a_list = a.splitlines(keepends=True)
+    b_list = b.splitlines(keepends=True)
+
+    with _patch_difflib_sequence_matcher_init():
+        # Set `n` to `sys.maxsize` to show entire graph when there is a diff.
+        diff = "".join(difflib.unified_diff(a_list, b_list, n=sys.maxsize))
+
+    if not diff:
+        return "<no diff>"
+    return diff
+
+
+@_beartype.beartype
+def _transform_diagnose_call_message_formatter(
+    run: Callable,
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+) -> str:
+    # TODO(bowbao): Update signature to varargs and varkwargs to avoid manual unpacking.
+    assert len(args) >= 1 and isinstance(args[0], Transform)
+    transform = args[0]
+
+    return f"Running {transform.__class__.__name__} pass."
+
+
+def fx_graph_tabular(graph: torch.fx.Graph) -> str:
+    """Return the Graph nodes in tabular format. Equivalent to stdout of `graph.print_tabular()`.
+
+    Args:
+        graph: The Graph to print.
+
+    Returns:
+        The Graph printed in a tabular format.
+    """
+    f = io.StringIO()
+    with contextlib.redirect_stdout(f):
+        graph.print_tabular()
+    return f.getvalue()
+
+
+class Transform(abc.ABC):
+    """Base class for FX graph transformations to be used by FX-ONNX exporter.
+
+    This class provides builtin support for transformation recording using the diagnostics system.
+
+    TODO(bowbao): Add more overrideable methods in call hierarchy
+    Methods in the Transform class can be overridden to customize the behavior of the
+    transform. The following methods can be overridden::
+
+        _run()
+            +-- run_node()
+                +-- placeholder()
+                +-- get_attr()
+                +-- call_function()
+                +-- call_method()
+                +-- call_module()
+                +-- output()
+
+    The granularity of overriding is up to the user. And it affects the granularity of
+    the diagnostics information. For example, if `_run()` is overridden, the
+    diagnostics information will only contain graph level transformation. Instead,
+    if `call_function()` is overridden, the diagnostics information will additionally
+    contain the node level information of `call_function()`.
+
+    Example: TODO(bowbao): Fill example once more overrideable methods are added.
+    """
+
+    """The module to be transformed."""
+    module: torch.fx.GraphModule
+
+    def __init__(self, module: torch.fx.GraphModule):
+        """Initialize the transform.
+
+        Args:
+            module: The module to be transformed.
+        """
+        self.module = module
+
+    @abc.abstractmethod
+    def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
+        ...
+
+    @diagnostics.diagnose_call(
```

Comment:
Thoughts on creating a hook on `run` to allow injecting `diagnostics.diagnose_call`?

Reply:
Could you share a snippet maybe, what is the main benefit?

Reply:
Just a thought. May not be a good idea now that I think about it